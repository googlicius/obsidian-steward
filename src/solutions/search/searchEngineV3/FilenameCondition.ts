import { Condition, ConditionResult } from './Condition';
import { ParsedRegexPattern } from '../types';
import { similarity } from 'src/utils/similarity';

const SIMILARITY_THRESHOLD = 0.7;

/**
 * Condition for filtering by filename.
 */
export class FilenameCondition extends Condition {
  constructor(private names: string[]) {
    super();
  }

  /**
   * Parse a regex pattern string and determine the search type and original name
   * Handles patterns like ^name$ (exact), ^name (startsWith), or name (contains)
   */
  private parseRegexPattern(pattern: string): ParsedRegexPattern {
    // Check if pattern is for exact match: ^name$
    if (pattern.startsWith('^') && pattern.endsWith('$')) {
      return {
        originalName: pattern.slice(1, -1),
        searchType: 'exact',
      };
    }

    // Check if pattern is for starts with: ^name
    if (pattern.startsWith('^')) {
      return {
        originalName: pattern.slice(1),
        searchType: 'startsWith',
      };
    }

    // Default case is contains: name
    return {
      originalName: pattern,
      searchType: 'contains',
    };
  }

  async evaluate() {
    const result = new Map<number, ConditionResult>();

    for (const name of this.names) {
      const terms = this.context.nameTokenizer.getUniqueTerms(name);
      if (terms.length === 0) continue;

      // Get document IDs from term entries
      const termEntries = await this.context.documentStore.getTermsByValue(terms);

      // Extract unique document IDs from term entries
      const documentIds = [...new Set(termEntries.map(entry => entry.documentId))];

      // If no matching documents found, continue to next name
      if (documentIds.length === 0) continue;

      // Fetch the documents for these IDs
      const documents = await this.context.documentStore.getDocumentsByIds(documentIds);

      // Filter documents by similarity score
      for (const doc of documents) {
        const parsedName = this.parseRegexPattern(name);
        let score = 0;

        switch (parsedName.searchType) {
          case 'contains':
            if (doc.fileName.toLowerCase().includes(parsedName.originalName.toLowerCase())) {
              score = 1.0;
            }
            break;

          case 'exact':
            score = similarity(parsedName.originalName, doc.fileName);
            break;

          case 'startsWith':
            if (doc.fileName.toLowerCase().startsWith(parsedName.originalName.toLowerCase())) {
              score = 1.0;
            }
            break;

          default:
            break;
        }

        // If similarity is above threshold, add to matched documents with score
        if (score >= SIMILARITY_THRESHOLD) {
          result.set(doc.id as number, {
            document: doc,
            score: score,
          });
        }
      }
    }

    return result;
  }
}
