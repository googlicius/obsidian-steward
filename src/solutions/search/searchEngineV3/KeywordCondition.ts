import { getQuotedQuery } from 'src/utils/getQuotedQuery';
import { Condition, ConditionResult } from './Condition';
import { ExactPhraseMatch } from '../types';
import { IndexedDocument, IndexedTerm, TermSource } from 'src/database/SearchDatabase';

const TERM_MATCH_THRESHOLD = 0.7;

/**
 * Condition for keyword search.
 */
export class KeywordCondition extends Condition<IndexedDocument> {
  constructor(private keywords: string[]) {
    super();
  }

  /**
   * Check if a keyword is an exact phrase (wrapped in quotes)
   * @param keyword The keyword to check
   * @returns The exact phrase details or null if not an exact phrase
   */
  private checkForExactPhrase(keyword: string): ExactPhraseMatch | null {
    const quotedContent = getQuotedQuery(keyword);

    if (quotedContent) {
      const phrase = quotedContent;
      const tokens = this.context.contentTokenizer.getUniqueTerms(phrase);
      return {
        originalPhrase: phrase,
        tokens,
      };
    }

    return null;
  }

  /**
   * Get term entries that match the specified content terms
   */
  private async getTermEntriesForContent(terms: string[]): Promise<IndexedTerm[]> {
    // Pre-compute arrays for better performance
    // const documentIdArray = scopedDocuments.map(doc => doc.id as number);
    // const folderIdArray = folders.map(folder => folder.id as number);

    // Apply filtering
    return (
      this.context.documentStore.terms
        .where('term')
        .anyOf(terms)
        .and(item => item.source === TermSource.Content)
        // .and(item => this.isDocumentMatch(item.documentId, documentIdArray, scopedDocuments.length))
        // .and(item => this.isFolderMatch(item.folderId, folderIdArray, folders.length))
        .toArray()
    );
  }

  /**
   * Check if a document has terms appearing consecutively
   * @param termsMap Map of terms to their positions in the document
   * @param tokens The tokens to check for consecutive positions
   * @returns True if the terms appear consecutively, false otherwise
   */
  private hasConsecutiveTerms(termsMap: Map<string, number[]>, tokens: string[]): boolean {
    if (tokens.length <= 1) {
      return true; // Single token is always "consecutive"
    }

    // Convert position arrays to Sets for O(1) lookup instead of O(n)
    const positionSets = new Map<string, Set<number>>();
    for (const token of tokens) {
      const positions = termsMap.get(token);
      if (!positions || positions.length === 0) {
        return false; // If any token is missing, no consecutive sequence possible
      }
      positionSets.set(token, new Set(positions));
    }

    // Get all positions for the first token
    const firstTokenPositions = positionSets.get(tokens[0]);
    if (!firstTokenPositions) {
      return false;
    }

    // For each position of the first token, check if it starts a consecutive sequence
    for (const startPos of firstTokenPositions) {
      let isConsecutive = true;

      // Check if all other tokens appear in consecutive positions
      for (let i = 1; i < tokens.length; i++) {
        const expectedPos = startPos + i;
        const tokenPositions = positionSets.get(tokens[i]);
        if (!tokenPositions) {
          isConsecutive = false;
          break;
        }

        // If the expected position is not found, this is not a consecutive sequence
        if (!tokenPositions.has(expectedPos)) {
          isConsecutive = false;
          break;
        }
      }

      // If we found a consecutive sequence, return true
      if (isConsecutive) {
        return true;
      }
    }

    // No consecutive sequence found
    return false;
  }

  /**
   * Handle exact phrase matching
   * @param exactPhrase The exact phrase to match
   * @param documentsMap Map to store matched documents
   */
  private async handleExactPhraseMatch(
    exactPhrase: ExactPhraseMatch,
    documentsMap: Map<number, ConditionResult<IndexedDocument>>
  ): Promise<void> {
    const { originalPhrase, tokens } = exactPhrase;

    if (tokens.length === 0) {
      return;
    }

    // Get term entries for all tokens in the phrase
    const termEntries = await this.getTermEntriesForContent(tokens);

    if (termEntries.length === 0) {
      return;
    }

    // Group by document ID
    const documentTermsMap = new Map<number, Map<string, number[]>>();

    // For each term entry, store the positions by document and term
    for (const entry of termEntries) {
      const { documentId, term, positions } = entry;

      if (!documentTermsMap.has(documentId)) {
        documentTermsMap.set(documentId, new Map());
      }

      const termsMap = documentTermsMap.get(documentId);
      if (!termsMap?.has(term)) {
        termsMap?.set(term, []);
      }

      termsMap?.get(term)?.push(...positions);
    }

    // Get documents that have all the tokens
    const potentialDocIds = Array.from(documentTermsMap.keys()).filter(docId => {
      const termsMap = documentTermsMap.get(docId);
      // Check if document has all tokens from the phrase
      return tokens.every(token => termsMap?.has(token));
    });

    if (potentialDocIds.length === 0) {
      return;
    }

    // For each potential document, check if it contains the exact phrase
    const exactMatchDocIds: number[] = [];

    for (const docId of potentialDocIds) {
      const termsMap = documentTermsMap.get(docId);
      if (!termsMap) continue;

      // Check for consecutive positions
      if (this.hasConsecutiveTerms(termsMap, tokens)) {
        exactMatchDocIds.push(docId);
      }
    }

    if (exactMatchDocIds.length === 0) {
      return;
    }

    // Fetch documents for the exact match IDs
    const exactMatchDocs = await this.context.documentStore.getDocumentsByIds(exactMatchDocIds);

    // Merge into overall result map
    for (const doc of exactMatchDocs) {
      const docId = doc.id as number;

      if (documentsMap.has(docId)) {
        // Document already exists, update score and keywords matched
        const existingResult = documentsMap.get(docId);
        if (existingResult) {
          existingResult.score += 10.0; // High score for exact matches
          existingResult.keywordsMatched = [
            ...(existingResult.keywordsMatched || []),
            originalPhrase,
          ];
        }
      } else {
        // New document, add to map
        documentsMap.set(docId, {
          document: doc,
          score: 10.0, // High score for exact matches
          keywordsMatched: [originalPhrase],
        });
      }
    }
  }

  /**
   * Group term entries by document
   */
  private groupTermEntriesByDocument(termEntries: IndexedTerm[]): Map<number, Set<string>> {
    const documentTermMap = new Map<number, Set<string>>();

    for (const entry of termEntries) {
      const { documentId, term } = entry;

      if (!documentTermMap.has(documentId)) {
        documentTermMap.set(documentId, new Set());
      }

      documentTermMap.get(documentId)?.add(term);
    }

    return documentTermMap;
  }

  /**
   * Get document IDs that meet the term match threshold
   */
  private getQualifiedDocumentIds(
    documentTermMap: Map<number, Set<string>>,
    terms: string[]
  ): number[] {
    const qualifiedIds: number[] = [];

    for (const [documentId, docTermsSet] of documentTermMap.entries()) {
      if (docTermsSet.size / terms.length >= TERM_MATCH_THRESHOLD) {
        qualifiedIds.push(documentId);
      }
    }

    return qualifiedIds;
  }

  async evaluate() {
    const documentsMap = new Map<number, ConditionResult<IndexedDocument>>();

    for (const keyword of this.keywords) {
      // Check if this is an exact phrase match
      const exactPhrase = this.checkForExactPhrase(keyword);

      if (exactPhrase) {
        // Handle exact phrase matching
        await this.handleExactPhraseMatch(exactPhrase, documentsMap);
        continue;
      }

      // Regular keyword matching (existing code)
      const terms = this.context.contentTokenizer.getUniqueTerms(keyword);
      if (terms.length === 0) {
        continue;
      }

      const termEntries = await this.getTermEntriesForContent(terms);
      const documentTermMap = this.groupTermEntriesByDocument(termEntries);

      // Get qualified document IDs for this keyword
      const keywordDocIds = this.getQualifiedDocumentIds(documentTermMap, terms);

      if (keywordDocIds.length === 0) {
        continue;
      }

      // Fetch documents for these IDs
      const keywordDocuments = await this.context.documentStore.getDocumentsByIds(keywordDocIds);

      // Calculate scores for this keyword using pre-tokenized terms
      const scoredDocumentsMap = await this.context.scoring.calculateDocumentScores(
        keywordDocuments,
        terms
      );

      // Filter out documents with 0 proximity bonus and merge into overall result map
      for (const [docId, scoredDoc] of scoredDocumentsMap.entries()) {
        // Skip documents with 0 proximity bonus
        if (scoredDoc.proximityBonus === 0) {
          continue;
        }

        if (documentsMap.has(docId)) {
          const existingResult = documentsMap.get(docId);
          if (existingResult) {
            existingResult.score += scoredDoc.score;
            existingResult.keywordsMatched = [...(existingResult.keywordsMatched || []), keyword];
          }
        } else {
          documentsMap.set(docId, {
            document: scoredDoc.document,
            score: scoredDoc.score,
            keywordsMatched: [keyword],
          });
        }
      }
    }

    return documentsMap;
  }
}
