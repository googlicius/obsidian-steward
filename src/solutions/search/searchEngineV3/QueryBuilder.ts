import { AndCondition } from './AndCondition';
import { Condition } from './Condition';
import { OrCondition } from './OrCondition';

/**
 * Generic QueryBuilder that allows adding any condition without business-specific methods.
 * Dependencies are provided through SearchContext to all conditions.
 */
export class QueryBuilder {
  private rootCondition: Condition;

  constructor() {
    this.rootCondition = new AndCondition();
  }

  /**
   * Add a condition with AND logic
   */
  and(condition: Condition): this {
    if (this.rootCondition instanceof AndCondition) {
      this.rootCondition = new AndCondition(...this.rootCondition['conditions'], condition);
    } else {
      this.rootCondition = new AndCondition(this.rootCondition, condition);
    }
    return this;
  }

  /**
   * Add a condition with OR logic
   */
  or(condition: Condition): this {
    if (this.rootCondition instanceof OrCondition) {
      this.rootCondition = new OrCondition(...this.rootCondition['conditions'], condition);
    } else {
      this.rootCondition = new OrCondition(this.rootCondition, condition);
    }
    return this;
  }

  build(): Condition {
    return this.rootCondition;
  }
}
