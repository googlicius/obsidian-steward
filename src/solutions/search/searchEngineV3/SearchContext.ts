import { DocumentStore } from '../documentStore';
import { Tokenizer } from '../tokenizer';
import { Scoring } from '../scoring';

/**
 * Context object that provides shared dependencies to Condition classes.
 * This eliminates the need to pass individual dependencies to each condition.
 */
export interface SearchContext {
  documentStore: DocumentStore;
  nameTokenizer: Tokenizer;
  contentTokenizer: Tokenizer;
  scoring: Scoring;
}
