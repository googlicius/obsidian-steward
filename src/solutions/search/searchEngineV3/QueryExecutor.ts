import { IndexedDocument } from 'src/database/SearchDatabase';
import { Condition } from './Condition';
import { logger } from 'src/utils/logger';
import { SearchContext } from './SearchContext';

/**
 * Interface for query results (array of matching documents with optional metadata)
 */
export interface QueryResult {
  documents: IndexedDocument[];
  count: number;
}

/**
 * Updated Executor class (uses DocumentStore for fetching).
 */
export class QueryExecutor {
  constructor(private context: SearchContext) {}

  async execute(condition: Condition): Promise<QueryResult> {
    try {
      const resultMap = await condition.injectContext(this.context).evaluate();
      if (resultMap.size === 0) return { documents: [], count: 0 };

      // Convert map to array and sort by score descending
      const sortedDocuments = Array.from(resultMap.entries())
        .sort((a, b) => b[1].score - a[1].score) // Higher score first; adjust if needed
        .map(entry => entry[1].document);

      return { documents: sortedDocuments, count: sortedDocuments.length };
    } catch (error) {
      logger.error('Query execution failed:', error);
      return { documents: [], count: 0 };
    }
  }
}
