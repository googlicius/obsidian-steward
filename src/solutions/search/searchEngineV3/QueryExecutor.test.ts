import { FilenameCondition } from './FilenameCondition';
import { KeywordCondition } from './KeywordCondition';
import { QueryBuilder } from './QueryBuilder';
import { QueryExecutor } from './QueryExecutor';
import { SearchContext } from './SearchContext';

describe('QueryExecutor', () => {
  it('should inject context to all conditions', async () => {
    // Create mock conditions with spy on evaluate method
    const filenameCondition = new FilenameCondition(['cat']);
    const keywordCondition = new KeywordCondition(['black cat']);

    // Mock the evaluate methods to return empty results
    const filenameEvaluateSpy = jest
      .spyOn(filenameCondition, 'evaluate')
      .mockResolvedValue(new Map());
    const keywordEvaluateSpy = jest
      .spyOn(keywordCondition, 'evaluate')
      .mockResolvedValue(new Map());

    const queryBuilder = new QueryBuilder();
    queryBuilder.and(filenameCondition);
    queryBuilder.and(keywordCondition);
    const condition = queryBuilder.build();

    const searchContext = {
      documentStore: {},
    } as SearchContext;

    const queryExecutor = new QueryExecutor(searchContext);

    await queryExecutor.execute(condition);

    expect(filenameEvaluateSpy).toHaveBeenCalled();
    expect(keywordEvaluateSpy).toHaveBeenCalled();
    expect(condition).toMatchSnapshot();
  });

  it('should return intersected results when using AND condition', async () => {
    const filenameCondition = new FilenameCondition(['cat']);
    const keywordCondition = new KeywordCondition(['black cat']);

    // Mock the evaluate methods to return specific results
    const filenameResults = new Map();
    filenameResults.set(1, { document: { id: 1, name: 'cat.md' }, score: 0.8 });
    filenameResults.set(2, { document: { id: 2, name: 'cat_photo.md' }, score: 0.7 });
    filenameResults.set(3, { document: { id: 3, name: 'another_cat.md' }, score: 0.6 });

    const keywordResults = new Map();
    keywordResults.set(1, { document: { id: 1, name: 'cat.md' }, score: 0.9 });
    keywordResults.set(3, { document: { id: 3, name: 'another_cat.md' }, score: 0.5 });
    keywordResults.set(4, { document: { id: 4, name: 'black_kitten.md' }, score: 0.4 });

    jest.spyOn(filenameCondition, 'evaluate').mockResolvedValue(filenameResults);
    jest.spyOn(keywordCondition, 'evaluate').mockResolvedValue(keywordResults);

    const queryBuilder = new QueryBuilder();
    queryBuilder.and(filenameCondition).and(keywordCondition);
    const condition = queryBuilder.build();

    const searchContext = {} as SearchContext;

    const queryExecutor = new QueryExecutor(searchContext);
    const result = await queryExecutor.execute(condition);

    expect(result).toMatchObject({
      conditionResults: [
        {
          document: {
            id: 1,
            name: 'cat.md',
          },
          keywordsMatched: [],
          score: 1.7000000000000002,
        },
        {
          document: {
            id: 3,
            name: 'another_cat.md',
          },
          keywordsMatched: [],
          score: 1.1,
        },
      ],
      count: 2,
    });
  });

  it('should return unioned results when using OR condition', async () => {
    const filenameCondition = new FilenameCondition(['cat']);
    const keywordCondition = new KeywordCondition(['black cat']);

    // Mock the evaluate methods to return specific results
    const filenameResults = new Map();
    filenameResults.set(1, { document: { id: 1, name: 'cat.md' }, score: 0.8 });
    filenameResults.set(2, { document: { id: 2, name: 'cat_photo.md' }, score: 0.7 });
    filenameResults.set(3, { document: { id: 3, name: 'another_cat.md' }, score: 0.6 });

    const keywordResults = new Map();
    keywordResults.set(1, { document: { id: 1, name: 'cat.md' }, score: 0.9 });
    keywordResults.set(3, { document: { id: 3, name: 'another_cat.md' }, score: 0.5 });
    keywordResults.set(4, { document: { id: 4, name: 'black_kitten.md' }, score: 0.4 });

    jest.spyOn(filenameCondition, 'evaluate').mockResolvedValue(filenameResults);
    jest.spyOn(keywordCondition, 'evaluate').mockResolvedValue(keywordResults);

    const queryBuilder = new QueryBuilder();
    queryBuilder.or(filenameCondition).or(keywordCondition);
    const condition = queryBuilder.build();

    const searchContext = {} as SearchContext;

    const queryExecutor = new QueryExecutor(searchContext);
    const result = await queryExecutor.execute(condition);

    expect(result).toMatchObject({
      conditionResults: [
        {
          document: {
            id: 1,
            name: 'cat.md',
          },
          score: 1.7000000000000002,
        },
        {
          document: {
            id: 3,
            name: 'another_cat.md',
          },
          score: 1.1,
        },
        {
          document: {
            id: 2,
            name: 'cat_photo.md',
          },
          score: 0.7,
        },
        {
          document: {
            id: 4,
            name: 'black_kitten.md',
          },
          score: 0.4,
        },
      ],
      count: 4,
    });
  });
});
