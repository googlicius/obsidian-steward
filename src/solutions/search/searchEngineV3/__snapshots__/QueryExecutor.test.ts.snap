// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryExecutor should inject context to all conditions 1`] = `
AndCondition {
  "conditions": [
    FilenameCondition {
      "context": {
        "documentStore": {},
      },
      "evaluate": [MockFunction] {
        "calls": [
          [],
        ],
        "results": [
          {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "names": [
        "cat",
      ],
    },
    KeywordCondition {
      "context": {
        "documentStore": {},
      },
      "evaluate": [MockFunction] {
        "calls": [
          [],
        ],
        "results": [
          {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "keywords": [
        "black cat",
      ],
    },
  ],
  "context": {
    "documentStore": {},
  },
}
`;
